p2p/src/sync/queue.rs

```diff
@@ -0,0 +1,707 @@
+            resolved.0.into_iter().for_each(|descendant| {
+                self.lookup.remove(descendant.get_id());
+                debug_assert!(!self.export.contains_key(descendant.get_id()));
+                self.queue(descendant);
+            });
+        }
+
+        Ok(ImportQueueState::Queued)
+    }
+
+    /// Try to queue element to the import queue and if it resolves a dependency,
+    /// return [`ImportQueueState::Resolved`] instead which indicates to the caller
+    /// they can now try and fetch all the data from the queue.
+    pub fn try_queue(&mut self, data: &T) -> error::Result<ImportQueueState> {
+        let prev_id = data.get_prev_id().ok_or(P2pError::InvalidData)?;
+
+        // dependency has been resolved if the export table contains an entry with this
+        // id but the lookup table doesn't and the entry's parent is not in the export table
+        if self.export.contains_key(data.get_id())
+            && !self.lookup.contains_key(&prev_id)
+            && !self.export.contains_key(&prev_id)
+        {
+            return Ok(ImportQueueState::Resolved);
+        }
+
+        self.queue(data.clone())
+    }
+
+    /// Add element to the import queue
+    pub fn queue(&mut self, data: T) -> error::Result<ImportQueueState> {
+        let prev_id = data.get_prev_id().ok_or(P2pError::InvalidData)?;
+        let id = *data.get_id();
+
+        if self.lookup.contains_key(data.get_id()) {
+            return Ok(ImportQueueState::Duplicate);
+        }
+
+        match self.export.get_mut(&prev_id) {
+            Some(ancestor) => {
+                self.lookup.insert(id, (prev_id, 0));
+                ancestor.queue(data, 0);
+            }
+            None => match self.lookup.get(&prev_id) {
+                Some(info) => {
+                    let (ancestor, idx) = (info.0, info.1 + 1);
+                    self.lookup.insert(id, (ancestor, idx));
+                    self.export.get_mut(&ancestor).ok_or(P2pError::InvalidData)?.queue(data, idx);
+                }
+                None => {
+                    self.lookup.insert(id, (prev_id, 0));
+                    self.export.insert(prev_id, OrderedData(vec![vec![data]]));
+                }
+            },
+        }
+
+        self.resolve_deps(&id)
+    }
+
+    /// Get queued descendants
+    pub fn drain_with_id(&mut self, data: &T::Id) -> Option<QueuedData<T>> {
+        if let Some(exported) = self.export.remove(data) {
+            let resolved = QueuedData(exported.0.into_iter().flatten().collect());
+
+            // remove all export and lookup references that are no longer needed
+            resolved.0.iter().for_each(|descendant| {
+                self.lookup.remove(descendant.get_id()).map_or_else(|| (), |_| ())
+            });
+            self.lookup.remove(data);
+
+            return Some(resolved);
+        }
+
+        None
+    }
+
+    /// Get all non-orphan chains from the import queue and clear all unresolved data
```

Comment:
Been wondering whether this is about orphan blocks... take a look at the OrphansBlockPool container in Consensus. You may have reinvented the wheel.

Reply:
I think it's different for two reasons. Firstly, all of the blocks have a block source different than `BlockSource::Local` so in `process_block()` error would be returned which is not something we want. Secondly the blocks are "orphan" only for a very brief moment in time and only because we download multiple blocks simultaneously and they may be received out of order. If we were to download them in order and one by one, there would be only valid blocks and actual orphan blocks in which case we could just directly import them to the chainstate. Btw, Substrate also uses an import queue for the block syncing, probably to solve this exact same problem.